query_optimizer/rules/UnnestSubqueries.cpp

```diff
@@ -0,0 +1,720 @@
+    if (new_input_expression != input_expression && !has_changed_expression) {
+      has_changed_expression = true;
+    }
+    new_input_expressions.emplace_back(new_input_expression);
+  }
+
+  if (has_changed_expression) {
+    // If there are correlated subquery expressions, add logical joins.
+    if (!correlated_query_info_vec.empty()) {
+      L::LogicalPtr new_child;
+
+      if (new_input_expressions[0] == nullptr
+          && node->getLogicalType() == L::LogicalType::kNestedLoopsJoin) {
+        new_child = node;
+      } else {
+        DCHECK_EQ(node->children().size(), 1u);
+        new_child = node->children()[0];
+      }
+
+      for (CorrelatedQueryInfo &correlated_query_info : correlated_query_info_vec) {
+        DCHECK(!correlated_query_info.probe_join_attributes.empty());
+        if (correlated_query_info.query_type_ == CorrelatedQueryInfo::kHashJoin) {
+          DCHECK(correlated_query_info.non_hash_join_predicates.empty())
+              << correlated_query_info.non_hash_join_predicates[0]->toString();
+          new_child = L::HashJoin::Create(new_child,
+                                          correlated_query_info.correlated_query,
+                                          correlated_query_info.probe_join_attributes,
+                                          correlated_query_info.build_join_attributes,
+                                          nullptr, /* residual_predicate */
+                                          L::HashJoin::kInnerJoin);
+        } else {
+          E::PredicatePtr filter_predicate;
+          if (correlated_query_info.non_hash_join_predicates.size() > 1u) {
+            filter_predicate = E::LogicalAnd::Create(correlated_query_info.non_hash_join_predicates);
+          } else if (!correlated_query_info.non_hash_join_predicates.empty()) {
+            filter_predicate = correlated_query_info.non_hash_join_predicates[0];
+          }
+
+          L::HashJoin::JoinType join_type =
+              (correlated_query_info.query_type_ == CorrelatedQueryInfo::kSemiJoin)
+                  ? L::HashJoin::kLeftSemiJoin
+                  : L::HashJoin::kLeftAntiJoin;
+          new_child = L::HashJoin::Create(new_child,
+                                          correlated_query_info.correlated_query,
+                                          correlated_query_info.probe_join_attributes,
+                                          correlated_query_info.build_join_attributes,
+                                          filter_predicate,
+                                          join_type);
+        }
+      }
+
+      // Special case for filter and nested loops join which may contain EXISTS and IN.
+      if (new_input_expressions[0] == nullptr) {
+        DCHECK_EQ(new_input_expressions.size(), 1u);
+        DCHECK(node->getLogicalType() == L::LogicalType::kFilter
+               || node->getLogicalType() == L::LogicalType::kNestedLoopsJoin) << node->toString();
```

Comment:
As a convention in our codebase, we put `&&` or `||` at the end of line. So it looks like

```
  DCHECK(node->getLogicalType() == L::LogicalType::kFilter ||
         node->getLogicalType() == L::LogicalType::kNestedLoopsJoin) << node->toString();
```